builder/builder-next/controller.go

```diff
@@ -36,12 +40,108 @@ import (
 	"github.com/moby/buildkit/util/archutil"
 	"github.com/moby/buildkit/util/entitlements"
 	"github.com/moby/buildkit/util/leaseutil"
+	"github.com/moby/buildkit/util/network/cniprovider"
+	"github.com/moby/buildkit/util/network/netproviders"
 	"github.com/moby/buildkit/worker"
+	"github.com/moby/buildkit/worker/containerd"
 	"github.com/pkg/errors"
 	bolt "go.etcd.io/bbolt"
 )
 
-func newController(rt http.RoundTripper, opt Opt) (*control.Controller, error) {
+func newController(rt http.RoundTripper, opt Opt) (*mobycontrol.Controller, error) {
+	if opt.UseSnapshotter {
+		return newSnapshotterController(rt, opt)
+	}
+	return newGrapDriverController(rt, opt)
+}
+
+func newSnapshotterController(rt http.RoundTripper, opt Opt) (*mobycontrol.Controller, error) {
+	if err := os.MkdirAll(opt.Root, 0711); err != nil {
+		return nil, err
+	}
+
+	dist := opt.Dist
+
+	cacheStorage, err := bboltcachestorage.NewStore(filepath.Join(opt.Root, "cache.db"))
+	if err != nil {
+		return nil, err
+	}
+
+	nc := netproviders.Opt{
+		Mode: "auto",
+		CNI: cniprovider.Opt{
+			Root:       opt.Root,
+			ConfigPath: "/etc/buildkit/cni.json",
```

Comment:
Are we going to allow this?
Also is this the right place for configuring dockerd?

Reply:
Good one. Not sure if this is needed for dockerd, but perhaps @tonistiigi has thoughts on that.

We should at least look at the config. I agree it's somewhat odd for `dockerd` to take configurations from `/etc/buildkit`. Perhaps this should be part of either `daemon.json` or a file in dockerd's config location (unfortunately, we currently only have a `--config-file` flag, not a `--config-dir`)